File: internal/core/topic_fold.go
// Fiscal convention: monetary values in the Drenyra ecosystem are int64 cents;
// no float is ever used for money. This module is the canonical topic-key fold
// (sdd-060-tenant-cli, FR-TEN-3): a PURE normalization that maps a raw topic
// key to its folded form for drift comparison, plus the pure drift-group
// shapes. It joins the Go↔TS golden parity mechanism (config golden_parity).
// No monetary field exists anywhere in this file.
package core

import "strings"

// FoldTopicKey normalizes a raw topic key for drift comparison:
//
//  1. Unicode lower-case (strings.ToLower);
//  2. every rune that is not a letter or digit becomes a SINGLE SPACE
//     separator (punctuation, slashes, hyphens, dots, marks — all fold to the
//     same separator, preserving word boundaries);
//  3. whitespace runs (real or punctuation-derived) are collapsed to one space
//     and the result is trimmed.
//
// Examples: "rule/IGV credit" and "rule/igv-credit" both fold to
// "rule igv credit" — formatting drift collides, while word boundaries
// survive. Accent folding is EXPLICITLY out of scope (the fold is conservative
// so genuinely distinct keys never collide). The result is deterministic and
// pure. Two chains drift when their folded keys are equal but their raw keys
// differ under the same exact scope tuple.
func FoldTopicKey(raw string) string {
	var b strings.Builder
	b.Grow(len(raw))
	pendingSpace := false
	for _, r := range strings.ToLower(raw) {
		switch {
		case r == ' ' || r == '\t' || r == '\n' || r == '\r':
			pendingSpace = true
		case isFoldRune(r):
			if pendingSpace && b.Len() > 0 {
				b.WriteByte(' ')
			}
			pendingSpace = false
			b.WriteRune(r)
		default:
			// punctuation, slashes, hyphens, dots, marks — a separator space.
			pendingSpace = true
		}
	}
	return b.String()
}

// isFoldRune reports whether the rune survives the fold: letters and digits
// (Unicode-aware; anything else — punctuation, symbols, marks — is dropped).
func isFoldRune(r rune) bool {
	return (r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z') ||
		(r >= '0' && r <= '9') ||
		(r >= 0x00C0 && r <= 0x024F) || // Latin-1 supplement + extended (incl. accented letters, kept as-is)
		(r >= 0x0370 && r <= 0x03FF) // Greek
}

// DriftedChain is one non-canonical chain in a drift group.
type DriftedChain struct {
	TopicKey  string `json:"topicKey"`
	ChainSize int    `json:"chainSize"`
}

// DriftGroup is one collision group: chains whose folded topic keys are equal
// but raw keys differ under one tenant tuple and one fiscal period (Period is
// the exact period of the group; "" never occurs — empty period scans the whole
// tenant and groups per period). Canonical is the deterministic candidate (most
// observations, ties broken lexicographically).
type DriftGroup struct {
	Period             string         `json:"period"`
	Canonical          string         `json:"canonical"`
	CanonicalChainSize int            `json:"canonicalChainSize"`
	Drifted            []DriftedChain `json:"drifted"`
}

// ConsolidateReport is the `tenant consolidate` document (dry-run and apply).
type ConsolidateReport struct {
	RUC                string       `json:"ruc"`
	Period             string       `json:"period"`
	DryRun             bool         `json:"dryRun"`
	DriftGroups        []DriftGroup `json:"driftGroups"`
	TotalDriftGroups   int          `json:"totalDriftGroups"`
	TotalDriftedChains int          `json:"totalDriftedChains"`
}

File: core/topic-fold.ts
/**
 * Fiscal convention: monetary values in the Drenyra ecosystem are BigInt cents;
 * no float is ever used for money.
 *
 * Canonical topic-key fold (sdd-060-tenant-cli, FR-TEN-3): the TypeScript twin
 * of internal/core/topic_fold.go. PURE and deterministic; joins the Go↔TS
 * golden parity mechanism (config golden_parity) — the same
 * testdata/golden/topic-fold.json vectors run from both runtimes and must agree
 * byte-identically.
 *
 * Fold contract:
 *   1. lower-case;
 *   2. every rune that is not a letter or digit becomes a SINGLE SPACE
 *      separator (punctuation, slashes, hyphens, dots, marks all fold to the
 *      same separator, preserving word boundaries);
 *   3. whitespace runs (real or punctuation-derived) collapse to one space and
 *      the result is trimmed.
 *
 * Accent folding is explicitly out of scope (the fold is conservative so
 * genuinely distinct keys never collide). "rule/IGV credit" and
 * "rule/igv-credit" both fold to "rule igv credit".
 */

/** FoldTopicKey normalizes a raw topic key for drift comparison. */
export function foldTopicKey(raw: string): string {
	let out = "";
	let pendingSpace = false;
	for (const ch of raw.toLowerCase()) {
		if (ch === " " || ch === "\t" || ch === "\n" || ch === "\r") {
			pendingSpace = true;
		} else if (isFoldChar(ch)) {
			if (pendingSpace && out.length > 0) out += " ";
			pendingSpace = false;
			out += ch;
		} else {
			pendingSpace = true; // punctuation → separator space
		}
	}
	return out;
}

/** isFoldChar reports whether the character survives the fold (letters/digits). */
function isFoldChar(ch: string): boolean {
	const code = ch.codePointAt(0) ?? 0;
	return (
		(ch >= "a" && ch <= "z") ||
		(ch >= "A" && ch <= "Z") ||
		(ch >= "0" && ch <= "9") ||
		(code >= 0x00c0 && code <= 0x024f) || // Latin-1 supplement + extended
		(code >= 0x0370 && code <= 0x03ff) // Greek
	);
}
